### PR TITLE
feat(Auth): add custom endpoint

### PR DESCRIPTION
**Pull request by banji180**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
Read and validate endpoint param from the congnitouserpool section of the json configuration file

*How did you test these changes?*
(Please add a line here how the changes were tested)

- [x] Added Unit Tests
- [ ] Added Integration Tests

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/UserPoolConfiguration.kt [hunk 1]
```diff
@@ -62,24 +64,48 @@ data class UserPoolConfiguration internal constructor(val builder: Builder) {
...
 
         fun region(region: String) = apply { this.region = region }
+        fun endpoint(endpoint: String) = apply { this.endpoint = validateEndpoint(endpoint) }
         fun poolId(poolId: String) = apply { this.poolId = poolId }
         fun appClientId(appClientId: String) = apply { this.appClientId = appClientId }
         fun appClientSecret(appClientSecret: String) = apply { this.appClientSecret = appClientSecret }
         fun build() = UserPoolConfiguration(this)
+
+        @Throws(AuthException::class)
+        private fun validateEndpoint(endpoint: String?): String? {
```
Comment (sktimalsina) [inline aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/UserPoolConfiguration.kt:90]: Please write a test for this with known endpoints that should work and those that should not.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/AuthorizationCognitoActions.kt [hunk 2]
```diff
@@ -104,10 +106,16 @@ object AuthorizationCognitoActions : AuthorizationActions {
             }
             else -> null
         }
+
+        val clientMetadata = mutableMapOf<String, String>()
+        amplifyCredential.getCognitoSession().userPoolTokens.value?.accessToken?.let(SessionHelper::getUsername)
```
Comment (sktimalsina) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/AuthorizationCognitoActions.kt:111]: 🤯

**Review by sktimalsina — CHANGES_REQUESTED** [review]

**Review by sktimalsina — CHANGES_REQUESTED** [review]
Is this only PR in the chain? I cannot see how this custom endpoint is used by client. It would be helpful to see some tests.

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/UserPoolConfiguration.kt (hunk 1 above)
Comment (banji180) [reply aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/UserPoolConfiguration.kt:90]: Yes just added them

**Review by sktimalsina — APPROVED** [review]
Please make sure that the error type and message are same across platforms

**Review by gpanshu — APPROVED** [review]

**Comment by div5yesh** [comment]
Need a follow up on this. The custom endpoint is not used anywhere. We need to pass this value to cognito clients during configuration.

**Pull request merged**